README.md
apple2pi
========

What is the Apple II Pi?
------------------------
Basically, the Apple II Pi is the integration of an Apple II with a Raspberry Pi (http://www.raspberrypi.org) to create a hybrid computer combining the input devices and storage mediums of the Apple with the CPU, GPU (graphical processing unit), USB, and network capabilities of the Broadcom BCM2835 SoC (think smartphone processor).  The concept is to create an updated version of the Apple II using some imagination, low-level drivers, off-the-shelf hardware, and a closely coupled communications channel; thus bringing semi-modern hardware and 32 bit *nix software to the Apple II platform.  The Apple II is running as a dedicated I/O processor for the Raspberry Pi under ProDOS.  The Raspberry Pi is running a version of Debian Linux: Raspbian.  Much like the PC Transporter card brought MS-DOS and the Z-80 card brought CP/M, the Apple II Pi brings Linux to the Apple II using the Apple’s input devices and the Raspberry Pi’s video output.  As such, knowledge and familiarity with Linux is required to get the most out of its environment.  Under Linux, the Apple II Pi can read and write the Apple’s storage devices (floppies/harddrives/CFFA) and also run the GSport Apple IIgs emulator (http://gsport.sourceforge.net).  Together, GSport and Apple II Pi provide an immersive environment providing access to most of the Apple II hardware you own plus an accelerated ~20MHz 65816 with up to 8 MB RAM, and all the disk images you can fit on the SD card.

Apple II client/server for Raspberry Pi
---------------------------------------

Apple II Pi works by connecting an Apple II to a Raspberry Pi using a RS232 serial connection.  In order to get the Raspberry Pi to talk RS232 from it's 3.3V GPIO serial port, you will need to build or buy a converter.  They are very cheap on eBay, so I would recommend going that route. Alternatively, you can use a USB serial port converter, but you will need to know its tty device name. To ensure you've hooked the converter up correctly, try loggin into the Raspberry Pi from another modern-ish computer.  Raspbian, the default Debian based Linux for the Raspberry Pi, opens up a login (getty) session on the serial port at 115.2K baud.  You will probably need a null modem or cross-over cable to login from another computer.  Once it all checks out, time to connect your Apple II.  All the 3.3V converters and USB serial ports I see have a DB-9 connector and many of the Apple II era connectors are DB-25 so you may need a DB-9 to DB-25 converter.

Installing and configuring the Apple II:  You will need an Apple //c or Apple ][, //e, IIgs w/ SuperSerial Card.  An Apple ][ requires the SHIFT key mod.  An Apple II Mouse is recommended for that full-on retro feel, but not required.  Download and install the A2PI.PO disk image onto a 5 1/4 floppy.  ADTPro would be the recommended tool for that operation although once you have the latest apple2pi version running, you can use the included dskwrite and dskread utilities for writing and reading ProDOS floppies.

Installing and configuring the Raspberry Pi:  Download the apple2pi project to your Raspberry Pi.  Enter the apple2pi/src directory.  Compile the daemon and tools with 'make' and copy the results to /usr/local/bin with 'sudo make install'.  To build the FUSE driver needed to mount ProDOS devices under Linux, you will need the libfuse-dev package installed.  Get this from apt-get, aptitude, or whichever package manager you like.  Build with 'make fusea2pi' and install with 'sudo make fuse-install'.

The following is no longer neessary as the install script carefully makes all the following adjustments automatically.  I left this here so you know what the script is doing.


>You will need to disable the Raspbain serial login by editing /etc/inittab and commenting out the line (probably at the very bottom) to look like:<br>
<code>
\#T0:23:respawn:/sbin/getty -L ttyAMA0 115200 vt100
</code>
<br>
You will also want to disable the console messages that go out to the serial port in the /boot/cmdline.txt file.  Remove the "console=" clause and the "kgdboc=" clause from the /boot/cmdline.txt file.  Mine looks like:<br>
<code>
dwc_otg.lpm_enable=0 console=tty1 root=/dev/mmcblk0p2 rootfstype=ext4 elevator=deadline rootwait
</code>

>If you are using an HDMI port to display video, skip the overscan settings.  This was for my monitor and your values may vary (a lot). I adjust the NTSC output so it fits nicely on my //c monitor, I edited the setting in /boot/config.txt such that:

><code>
overscan_left=26<br>
overscan_right=-8<br>
overscan_top=-8<br>
\#overscan_bottom=16<br>
</code>

>To run the a2pid daemon automatically at boot time, edit /etc/rc.local and add:

><code>
/usr/local/bin/a2pid --daemon
</code>

>right before the line:

><code>
exit 0
</code>

followed by rebooting the Raspberry Pi.

NOTE - For USB serial port users and non-Raspberry Pi owners:  This isn't actually tied to the Raspberry Pi in any way, except for the default serial port used to connect the Pi to the Apple II.  On most up-to-date Linux distributions, you should be able to build all the files.  To run the daemon on a specific serial port, just add it as a command line option i.e. a2pid --daemon /dev/ttyUSB0

Reboot the Apple II with the newly created floppy in the start-up drive.  If everything is configured correctly, you should be able to login to the Raspberry Pi with your Apple II keyboard.  If you have an Apple II Mouse, that should control the cursor in X, or in the console if you have gdm installed.

Using a2pi: The Apple //c and //e keyboards are pretty minimal compared to modern keyboards, and the Apple II Mouse only has one button.  In order to provide most of the funcitonality required of modern OSes, the Open-Apple and Closed-Apple keys are used as modifiers to enhance the keyboard and mouse.  On the keyboard, Open-Apple acts just like the Alt key.  The Closed-Apple key acts like a Fn key, changing the actual key codes.  Currently, the Closed-Apple key will modify the number keys 1-0 as funciton keys F1-F10 and the arrow keys as Left-Arrow=Home, Right-Arrow=End, Up-Arrow=PgUp, Down-Arrow=PgDn.  For the mouse, when you click the mouse button by itself, that is the left(default)-click.  Open-Apple along with the mouse button will return the right-click, and Closed-Apple along with the mouse button will return the middle-click.  a2pid can be run directly (not as a daemon) by leaving off the '--daemon' option.  Enabling printf's in the code allows one to watch the packets arrive and get processed when run from a network ssh session.

Theory of operation:  Apple II Pi works by running code on the Apple II and the Raspberry Pi, talking to each other with a simple protocol.  The Apple II basically appears to the Raspberry Pi as an external peripheral, not unlike a USB keyboard and mouse.  The Apple II floppy boots into ProDOS and runs a simple machine language program that scans the keyboard, and mouse if present, sending the events out the serial port to the Raspberry Pi.  It is a very simple protocol and the serial port is running at 115.2K baud, so it is fast and low overhead.  On the Raspberry Pi, a little daemon runs, waiting for packets on the serial port, converts the Apple II events into Linux compatible events, and inserts them into the input subsystem.  This daemon also has a socket interface (port 6551) that can be used to access the Apple II memory and execute arbitrary code.  Look at a2lib.c for implementation.

Enjoy,

Dave Schmenk


debian/changelog
a2pi (0.2.3-10) unstable; urgency=medium

  * Change PIROM signature to not conflict with Slinky RAM drivers

 -- David Schmenk <dschmenk@mail.com>  Wed, 12 May 2021 15:40:00 -0800

a2pi (0.2.3-5) unstable; urgency=medium

  * Flush writes to piDrives

 -- David Schmenk <dschmenk@mail.com>  Sun, 9 May 2021 12:28:00 -0800

a2pi (0.2.3-4) unstable; urgency=low

  * A2Pi 6.5 updates

 -- David Schmenk <dschmenk@mail.com>  Sun, 25 Apr 2021 13:44:30 -0800

a2pi (0.2.3-3) unstable; urgency=low

  * First Stretch release

 -- David Schmenk <dschmenk@mail.com>  Wed, 23 Aug 2017 13:44:30 -0800

a2pi (0.2.2-2) unstable; urgency=low

  * Change default serial port to serial0
  * Add plymouth splash screen

 -- David Schmenk <dschmenk@mail.com>  Thu, 08 Jun 2017 21:22:30 -0800

a2pi (0.2.2-0) unstable; urgency=low

  * First systemd compatibility

 -- David Schmenk <dschmenk@gmail.com>  Sat, 20 May 2017 15:15:30 -0800

a2pi (0.2.1-0) unstable; urgency=low

  * First Jessie package

 -- David Schmenk <dschmenk@gmail.com>  Tue, 17 Nov 2015 19:27:59 -0800

a2pi (0.2.0-2) unstable; urgency=low

  * Add support for Raspberry Pi version 2
  * Add support for A2Pi ROMM detect

 -- David Schmenk <dschmenk@gmail.com>  Thu, 09 Jul 2015 20:48:30 -0800

a2pi (0.1.8-8) unstable; urgency=low

  * Full ROM build
  * Add a2pi logo with transparent background
  * Only set GPCLK on default RPi UART
  * Stabilize request/pidrive interaction
  * Fix cancel from client connect & PIDRIVE

 -- David Schmenk <dschmenk@gmail.com>  Sat, 15 Feb 2014 13:15:00 -0800

a2pi (0.1.8-4) unstable; urgency=low

  * Fix FUSE create file function
  * Add VSDRIVE functionality
  * Fix recursive call request (FUSE mount w/ pidrive active)
  * Fix symlinks for A2VDx.PO updating
  * Add UTILS.PO image for comms programs and file manipulation
  * Run PIDRIVE from STARTUP

 -- David Schmenk <dschmenk@gmail.com>  Tue, 13 Feb 2014 15:31:23 -0800

a2pi (0.1.7-3) unstable; urgency=low

  * Adjust file permissions for FUSE driver to allow group access
  * Add options to allow root to access FUSE mount point (for GSport)
  * Update Apple ][+ keyboard support (Caps Lock)

 -- David Schmenk <dschmenk@gmail.com>  Fri, 06 Dec 2013 20:39:14 -0800

a2pi (0.1.6-1) unstable; urgency=low

  * One more attempt at mouse acceleration.  Better inital Apple client setup.

 -- David Schmenk <dschmenk@gmail.com>  Sun, 1 Dec 2013 18:59:17 -0800

a2pi (0.1.5-1) unstable; urgency=low

  * One more attempt at mouse acceleration, Apple ][/][+ support

 -- David Schmenk <dschmenk@gmail.com>  Mon, 16 Nov 2013 22:17:40 -0800

a2pi (0.1.4-2) unstable; urgency=low

  * Fix postinst to not comment out all T* lines

 -- David Schmenk <dschmenk@gmail.com>  Mon, 10 Nov 2013 04:48:20 -0800

a2pi (0.1.4-1) unstable; urgency=low

  * Change mouse accleration code from table lookup to squared value

 -- David Schmenk <dschmenk@gmail.com>  Mon, 4 Nov 2013 10:59:43 -0800

a2pi (0.1.3-1) unstable; urgency=low

  * Fix multi-thread locking in FUSE driver

 -- David Schmenk <dschmenk@gmail.com>  Tue, 8 Oct 2013 11:23:34 -0700

a2pi (0.1.2-1) unstable; urgency=low

  * Stop and start a2pid server before and after upgrade

 -- David Schmenk <dschmenk@gmail.com>  Mon, 7 Oct 2013 15:43:53 -0700

a2pi (0.1.1-3) unstable; urgency=low

  * Fix update to /boot/cmdline.txt - thanks Ivan Drucker!

 -- David Schmenk <dschmenk@gmail.com>  Sun, 6 Oct 2013 14:11:55 -0700

a2pi (0.1.1-2) unstable; urgency=low

  * Add user 'pi' to group 'fuse'

 -- David Schmenk <dschmenk@gmail.com>  Sun, 22 Sep 2013 15:16:00 -0700

a2pi (0.1.1-1) unstable; urgency=low

  * Clean up daemon restart code

 -- David Schmenk <dschmenk@gmail.com>  Sun, 22 Sep 2013 15:00:00 -0700

a2pi (0.1.0-4) unstable; urgency=low

  * Remove hard-coded path in a2mount

 -- David Schmenk <dschmenk@gmail.com>  Sun, 22 Sep 2013 14:32:30 -0700

a2pi (0.1.0-3) unstable; urgency=low

  * Fix for daemon start

 -- David Schmenk <dschmenk@gmail.com>  Sun, 22 Sep 2013 13:32:30 -0700

a2pi (0.1.0-2) unstable; urgency=low

  * Fix deb package postinst script

 -- David Schmenk <dschmenk@gmail.com>  Sun, 22 Sep 2013 13:02:32 -0700

a2pi (0.1.0-1) unstable; urgency=low

  * Initial release

 -- David Schmenk <dschmenk@gmail.com>  Thu, 19 Sep 2013 18:28:32 -0700


src/kbmap.h

int keycode[256] = {
    /*
     * normal scancode
     */
    MOD_CTRL | MOD_SHIFT | KEY_2,           // CTRL-@ code 00
    MOD_CTRL |             KEY_A,           // CTRL-A code 01
    MOD_CTRL |             KEY_B,           // CTRL-B code 02
    MOD_CTRL |             KEY_C,           // CTRL-C code 03
    MOD_CTRL |             KEY_D,           // CTRL-D code 04
    MOD_CTRL |             KEY_E,           // CTRL-E code 05
    MOD_CTRL |             KEY_F,           // CTRL-F code 06
    MOD_CTRL |             KEY_G,           // CTRL-G code 07
   			   KEY_LEFT,        // CTRL-H code 08
                           KEY_TAB,         // CTRL-I code 09
                           KEY_DOWN,        // CTRL-J code 0A
                           KEY_UP,          // CTRL-K code 0B
    MOD_CTRL |             KEY_L,           // CTRL-L code 0C
                           KEY_ENTER,       // CTRL-M code 0D
    MOD_CTRL |             KEY_N,           // CTRL-N code 0E
    MOD_CTRL |             KEY_O,           // CTRL-O code 0F
    MOD_CTRL |             KEY_P,           // CTRL-P code 10
    MOD_CTRL |             KEY_Q,           // CTRL-Q code 11
    MOD_CTRL |             KEY_R,           // CTRL-R code 12
    MOD_CTRL |             KEY_S,           // CTRL-S code 13
    MOD_CTRL |             KEY_T,           // CTRL-T code 14
                           KEY_RIGHT,       // CTRL-U code 15
    MOD_CTRL |             KEY_V,           // CTRL-V code 16
    MOD_CTRL |             KEY_W,           // CTRL-W code 17
    MOD_CTRL |             KEY_X,           // CTRL-X code 18
    MOD_CTRL |             KEY_Y,           // CTRL-Y code 19
    MOD_CTRL |             KEY_Z,           // CTRL-Z code 1A
                           KEY_ESC,         // ESCAPE code 1B
    MOD_CTRL |             KEY_BACKSLASH,   // CTRL-\ code 1C
    MOD_CTRL |             KEY_RIGHTBRACE,  // CTRL-] code 1D
    MOD_CTRL |             KEY_6,           // CTRL-6 code 1E
    MOD_CTRL |             KEY_MINUS,       // CTRL-- code 1F
                           KEY_SPACE,       // ' '    code 20
               MOD_SHIFT | KEY_1,           // !      code 21
               MOD_SHIFT | KEY_APOSTROPHE,  // "      code 22
               MOD_SHIFT | KEY_3,           // #      code 23
               MOD_SHIFT | KEY_4,           // $      code 24
               MOD_SHIFT | KEY_5,           // %      code 25
               MOD_SHIFT | KEY_7,           // &      code 26
                           KEY_APOSTROPHE,  // '      code 27
               MOD_SHIFT | KEY_9,           // (      code 28
               MOD_SHIFT | KEY_0,           // )      code 29
               MOD_SHIFT | KEY_8,           // *      code 2A
               MOD_SHIFT | KEY_EQUAL,       // +      code 2B
                           KEY_COMMA,       // ,      code 2C
                           KEY_MINUS,       // -      code 2D
                           KEY_DOT,         // .      code 2E
                           KEY_SLASH,       // /      code 2F
                           KEY_0,           // 0      code 30
                           KEY_1,           // 1      code 31
                           KEY_2,           // 2      code 32
                           KEY_3,           // 3      code 33
                           KEY_4,           // 4      code 34
                           KEY_5,           // 5      code 35
                           KEY_6,           // 6      code 36
                           KEY_7,           // 7      code 37
                           KEY_8,           // 8      code 38
                           KEY_9,           // 9      code 39
               MOD_SHIFT | KEY_SEMICOLON,   // :      code 3A
                           KEY_SEMICOLON,   // ;      code 3B
               MOD_SHIFT | KEY_COMMA,       // <      code 3C
                           KEY_EQUAL,       // =      code 3D
               MOD_SHIFT | KEY_DOT,         // >      code 3E
               MOD_SHIFT | KEY_SLASH,       // ?      code 3F
               MOD_SHIFT | KEY_2,           // @      code 40
               MOD_SHIFT | KEY_A,           // A      code 41
               MOD_SHIFT | KEY_B,           // B      code 42
               MOD_SHIFT | KEY_C,           // C      code 43
               MOD_SHIFT | KEY_D,           // D      code 44
               MOD_SHIFT | KEY_E,           // E      code 45
               MOD_SHIFT | KEY_F,           // F      code 46
               MOD_SHIFT | KEY_G,           // G      code 47
               MOD_SHIFT | KEY_H,           // H      code 48
               MOD_SHIFT | KEY_I,           // I      code 49
               MOD_SHIFT | KEY_J,           // J      code 4A
               MOD_SHIFT | KEY_K,           // K      code 4B
               MOD_SHIFT | KEY_L,           // L      code 4C
               MOD_SHIFT | KEY_M,           // M      code 4D
               MOD_SHIFT | KEY_N,           // N      code 4E
               MOD_SHIFT | KEY_O,           // O      code 4F
               MOD_SHIFT | KEY_P,           // P      code 50
               MOD_SHIFT | KEY_Q,           // Q      code 51
               MOD_SHIFT | KEY_R,           // R      code 52
               MOD_SHIFT | KEY_S,           // S      code 53
               MOD_SHIFT | KEY_T,           // T      code 54
               MOD_SHIFT | KEY_U,           // U      code 55
               MOD_SHIFT | KEY_V,           // V      code 56
               MOD_SHIFT | KEY_W,           // W      code 57
               MOD_SHIFT | KEY_X,           // X      code 58
               MOD_SHIFT | KEY_Y,           // Y      code 59
               MOD_SHIFT | KEY_Z,           // Z      code 5A
                           KEY_LEFTBRACE,   // [      code 5B
                           KEY_BACKSLASH,   // \      code 5C
                           KEY_RIGHTBRACE,  // ]      code 5D
               MOD_SHIFT | KEY_6,           // ^      code 5E
               MOD_SHIFT | KEY_MINUS,       // _      code 5F
                           KEY_GRAVE,       // `      code 60
                           KEY_A,           // a      code 61
                           KEY_B,           // b      code 62
                           KEY_C,           // c      code 63
                           KEY_D,           // d      code 64
                           KEY_E,           // e      code 65
                           KEY_F,           // f      code 66
                           KEY_G,           // g      code 67
                           KEY_H,           // h      code 68
                           KEY_I,           // i      code 69
                           KEY_J,           // j      code 6A
                           KEY_K,           // k      code 6B
                           KEY_L,           // l      code 6C
                           KEY_M,           // m      code 6D
                           KEY_N,           // n      code 6E
                           KEY_O,           // o      code 6F
                           KEY_P,           // p      code 70
                           KEY_Q,           // q      code 71
                           KEY_R,           // r      code 72
                           KEY_S,           // s      code 73
                           KEY_T,           // t      code 74
                           KEY_U,           // u      code 75
                           KEY_V,           // v      code 76
                           KEY_W,           // w      code 77
                           KEY_X,           // x      code 78
                           KEY_Y,           // y      code 79
                           KEY_Z,           // z      code 7A
               MOD_SHIFT | KEY_LEFTBRACE,   // {      code 7B
               MOD_SHIFT | KEY_BACKSLASH,   // |      code 7C
               MOD_SHIFT | KEY_RIGHTBRACE,  // }      code 7D
               MOD_SHIFT | KEY_GRAVE,       // ~      code 7E
                           KEY_BACKSPACE,   // BS     code 7F
    /*
     * w/ solid apple scancodes
     */
    MOD_CTRL | MOD_SHIFT | KEY_2,           // CTRL-@ code 00
    MOD_CTRL |             KEY_A,           // CTRL-A code 01
    MOD_CTRL |             KEY_B,           // CTRL-B code 02
    MOD_CTRL |             KEY_C,           // CTRL-C code 03
    MOD_CTRL |             KEY_D,           // CTRL-D code 04
    MOD_CTRL |             KEY_E,           // CTRL-E code 05
    MOD_CTRL |             KEY_F,           // CTRL-F code 06
    MOD_CTRL |             KEY_G,           // CTRL-G code 07
                           KEY_HOME,        // CTRL-H code 08
                           KEY_INSERT,      // CTRL-I code 09
                           KEY_PAGEDOWN,    // CTRL-J code 0A
                           KEY_PAGEUP,      // CTRL-K code 0B
    MOD_CTRL |             KEY_L,           // CTRL-L code 0C
                           KEY_LINEFEED,    // CTRL-M code 0D
    MOD_CTRL |             KEY_N,           // CTRL-N code 0E
    MOD_CTRL |             KEY_O,           // CTRL-O code 0F
    MOD_CTRL |             KEY_P,           // CTRL-P code 10
    MOD_CTRL |             KEY_Q,           // CTRL-Q code 11
    MOD_CTRL |             KEY_R,           // CTRL-R code 12
    MOD_CTRL |             KEY_S,           // CTRL-S code 13
    MOD_CTRL |             KEY_T,           // CTRL-T code 14
                           KEY_END,         // CTRL-U code 15
    MOD_CTRL |             KEY_V,           // CTRL-V code 16
    MOD_CTRL |             KEY_W,           // CTRL-W code 17
    MOD_CTRL |             KEY_X,           // CTRL-X code 18
    MOD_CTRL |             KEY_Y,           // CTRL-Y code 19
    MOD_CTRL |             KEY_Z,           // CTRL-Z code 1A
    MOD_CTRL |             KEY_ESC,         // ESCAPE code 1B
    MOD_CTRL |             KEY_BACKSLASH,   // CTRL-\ code 1C
    MOD_CTRL |             KEY_RIGHTBRACE,  // CTRL-] code 1D
    MOD_CTRL |             KEY_6,           // CTRL-6 code 1E
    MOD_CTRL |             KEY_MINUS,       // CTRL-- code 1F
                           KEY_SPACE,       // ' '    code 20
               MOD_SHIFT | KEY_F1,          // !      code 21
               MOD_SHIFT | KEY_APOSTROPHE,  // "      code 22
               MOD_SHIFT | KEY_F3,          // #      code 23
               MOD_SHIFT | KEY_F4,          // $      code 24
               MOD_SHIFT | KEY_F5,          // %      code 25
               MOD_SHIFT | KEY_F7,          // &      code 26
                           KEY_APOSTROPHE,  // '      code 27
               MOD_SHIFT | KEY_F9,          // (      code 28
               MOD_SHIFT | KEY_F10,         // )      code 29
               MOD_SHIFT | KEY_F8,          // *      code 2A
               MOD_SHIFT | KEY_F12,         // +      code 2B
                           KEY_COMMA,       // ,      code 2C
                           KEY_F11,         // -      code 2D
                           KEY_DOT,         // .      code 2E
                           KEY_SLASH,       // /      code 2F
                           KEY_F10,         // 0      code 30
                           KEY_F1,          // 1      code 31
                           KEY_F2,          // 2      code 32
                           KEY_F3,          // 3      code 33
                           KEY_F4,          // 4      code 34
                           KEY_F5,          // 5      code 35
                           KEY_F6,          // 6      code 36
                           KEY_F7,          // 7      code 37
                           KEY_F8,          // 8      code 38
                           KEY_F9,          // 9      code 39
               MOD_SHIFT | KEY_SEMICOLON,   // :      code 3A
                           KEY_SEMICOLON,   // ;      code 3B
               MOD_SHIFT | KEY_COMMA,       // <      code 3C
                           KEY_F12,         // =      code 3D
               MOD_SHIFT | KEY_DOT,         // >      code 3E
               MOD_SHIFT | KEY_SLASH,       // ?      code 3F
               MOD_SHIFT | KEY_F2,          // @      code 40
               MOD_SHIFT | KEY_A,           // A      code 41
               MOD_SHIFT | KEY_B,           // B      code 42
               MOD_SHIFT | KEY_C,           // C      code 43
               MOD_SHIFT | KEY_D,           // D      code 44
               MOD_SHIFT | KEY_E,           // E      code 45
               MOD_SHIFT | KEY_F,           // F      code 46
               MOD_SHIFT | KEY_G,           // G      code 47
               MOD_SHIFT | KEY_H,           // H      code 48
               MOD_SHIFT | KEY_I,           // I      code 49
               MOD_SHIFT | KEY_J,           // J      code 4A
               MOD_SHIFT | KEY_K,           // K      code 4B
               MOD_SHIFT | KEY_L,           // L      code 4C
               MOD_SHIFT | KEY_M,           // M      code 4D
               MOD_SHIFT | KEY_N,           // N      code 4E
               MOD_SHIFT | KEY_O,           // O      code 4F
               MOD_SHIFT | KEY_P,           // P      code 50
               MOD_SHIFT | KEY_Q,           // Q      code 51
               MOD_SHIFT | KEY_R,           // R      code 52
               MOD_SHIFT | KEY_S,           // S      code 53
               MOD_SHIFT | KEY_T,           // T      code 54
               MOD_SHIFT | KEY_U,           // U      code 55
               MOD_SHIFT | KEY_V,           // V      code 56
               MOD_SHIFT | KEY_W,           // W      code 57
               MOD_SHIFT | KEY_X,           // X      code 58
               MOD_SHIFT | KEY_Y,           // Y      code 59
               MOD_SHIFT | KEY_Z,           // Z      code 5A
                           KEY_LEFTBRACE,   // [      code 5B
                           KEY_BACKSLASH,   // \      code 5C
                           KEY_RIGHTBRACE,  // ]      code 5D
               MOD_SHIFT | KEY_F6,          // ^      code 5E
               MOD_SHIFT | KEY_F11,         // _      code 5F
                           KEY_GRAVE,       // `      code 60
                           KEY_A,           // a      code 61
                           KEY_B,           // b      code 62
                           KEY_C,           // c      code 63
                           KEY_D,           // d      code 64
                           KEY_E,           // e      code 65
                           KEY_F,           // f      code 66
                           KEY_G,           // g      code 67
                           KEY_H,           // h      code 68
                           KEY_I,           // i      code 69
                           KEY_J,           // j      code 6A
                           KEY_K,           // k      code 6B
                           KEY_L,           // l      code 6C
                           KEY_M,           // m      code 6D
                           KEY_N,           // n      code 6E
                           KEY_O,           // o      code 6F
                           KEY_P,           // p      code 70
                           KEY_Q,           // q      code 71
                           KEY_R,           // r      code 72
                           KEY_S,           // s      code 73
                           KEY_T,           // t      code 74
                           KEY_U,           // u      code 75
                           KEY_V,           // v      code 76
                           KEY_W,           // w      code 77
                           KEY_X,           // x      code 78
                           KEY_Y,           // y      code 79
                           KEY_Z,           // z      code 7A
               MOD_SHIFT | KEY_LEFTBRACE,   // {      code 7B
               MOD_SHIFT | KEY_BACKSLASH,   // |      code 7C
               MOD_SHIFT | KEY_RIGHTBRACE,  // }      code 7D
               MOD_SHIFT | KEY_GRAVE,       // ~      code 7E
                           KEY_DELETE       // DELETE code 7F
};
